component-api/src/main/java/org/talend/sdk/component/api/service/schema/DiscoverProcessorSchema.java

```diff
@@ -0,0 +1,48 @@
+/**
+ * Copyright (C) 2006-2022 Talend Inc. - www.talend.com
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.talend.sdk.component.api.service.schema;
+
+import static java.lang.annotation.ElementType.METHOD;
+import static java.lang.annotation.RetentionPolicy.RUNTIME;
+
+import java.lang.annotation.Retention;
+import java.lang.annotation.Target;
+
+import org.talend.sdk.component.api.meta.Documentation;
+import org.talend.sdk.component.api.meta.Partial;
+import org.talend.sdk.component.api.service.ActionType;
+
+/**
+ * Mark a method as returning a {@link org.talend.sdk.component.api.record.Schema} resulting from an input one
+ * and action of a processor for an outgoing branch.
+ */
+@Partial("See Schema description.")
+@ActionType(value = "processor_schema", expectedReturnedType = org.talend.sdk.component.api.record.Schema.class)
+@Target(METHOD)
+@Retention(RUNTIME)
+@Documentation("Mark a method as returning a Schema resulting from an input one and action of a processor for an outgoing branch. `value' name should match the connector's name.")
+public @interface DiscoverProcessorSchema {
```

Comment:
also please see the document :
https://talend.github.io/component-runtime/main/latest/component-define-processor-output.html
it allow mulitinput like : 
```
@ElementListener
public MyNewData map(@Input final MyData data, @Input("input2") final MyData2 data2) {
    return createNewData(data1, data2);
}
```
but from the jira desc, seems ignore this case, is right?

PS : for multi output like main/reject line or others, we have "branch" parameter.

Reply:
another question : 
how we locate the ui option in model for filling value? as from the design in the jira, can't find it. If no schema in model, how developer get it at runtime?

Reply:
PS : studio use this interface method to decide the allowed input and output lines for processor:
```
@ElementListener
    public Object next(@Input("in1") Record in1, @Input("in2") Record in2, @Output("out1") Record out1, @Output("out2") Record out2) {
        
    }
```

Reply:
> why not make it common for input and processor component? instead of the old way :
> 
> ```
> @Structure(type = Structure.Type.OUT, discoverSchema = "JDBCQueryDataSet")
> @Documentation("schema")
> private List<SchemaInfo> schema;//contains all info, but cloud part no convert now, and show this as a table widget in cloud.
> ```
> 
> ```
> @Structure(type = Structure.Type.OUT, discoverSchema = "JDBCQueryDataSet")
> @Documentation("schema")
> private List<String> schema;//no db column name/length/precision and so on, and show this as selection widget in cloud, and cloud ui have List<String> convert auto
> ```
> 
> And this no much meaning for jdbc connector at least:
> 
> ```
> type = Structure.Type.OUT
> ```
> 
> mean never use Structure.Type.IN

Your're right, I've already thought about that...
I thought about renaming the annotation as it could be used more generically and not only on processors  (usable by in/out connectors). Maybe stmg like `DiscoverSchemaForConnector`, `DiscoverSchemaExt`, `DiscoverSchemaXXX` 🤔

Reply:
> also please see the document : https://talend.github.io/component-runtime/main/latest/component-define-processor-output.html it allow mulitinput like :
> 
> ```
> @ElementListener
> public MyNewData map(@Input final MyData data, @Input("input2") final MyData2 data2) {
>     return createNewData(data1, data2);
> }
> ```
> 
> but from the jira desc, seems ignore this case, is right?
> 
> PS : for multi output like main/reject line or others, we have "branch" parameter.

Yes, this is voluntary dismissed for now.
Talked about this this week during some meetings...
Idea behind this signature is having a kind of tmap like and, in fact, it's not used yet by anyone!

Reply:
> 



> another question : how we locate the ui option in model for filling value? as from the design in the jira, can't find it. If no schema in model, how developer get it at runtime?

That's the goal of another issue in the backlog but not targeted to this particular issue.